lib/logstash/inputs/kafka.rb

```diff
@@ -133,31 +133,32 @@ def run(logstash_queue)
     @logger.info('Running kafka', :group_id => @group_id, :topic_id => @topic_id, :zk_connect => @zk_connect)
     begin
       @consumer_group.run(@consumer_threads,@kafka_client_queue)
-      begin
-        while true
-          event = @kafka_client_queue.pop
-          queue_event(event, logstash_queue)
-        end
-      rescue LogStash::ShutdownSignal
-        @logger.info('Kafka got shutdown signal')
-        @consumer_group.shutdown
+
+      while !stop?
+        event = @kafka_client_queue.pop
+        queue_event(event, logstash_queue)
       end
+
       until @kafka_client_queue.empty?
         queue_event(@kafka_client_queue.pop,logstash_queue)
       end
+
       @logger.info('Done running kafka input')
     rescue => e
       @logger.warn('kafka client threw exception, restarting',
                    :exception => e)
-      if @consumer_group.running?
-        @consumer_group.shutdown
-      end
+      stop
       sleep(Float(@consumer_restart_sleep_ms) * 1 / 1000)
```

Comment:
I'm wondering what happen here, you added the stop call, but then after it just go into an sleep? would it be possible the thread is waiting here? ... if so, I think it would be necessary to keep track of this so the stop call is able to wakeup the thread from the sleep. What do you think?